### PR TITLE
Initial array optimizations

### PR DESCRIPTION
**Pull request by Fletch153**

### CODE REVIEW

File: src/libraries/ArrayUtil.sol [hunk 1]
```diff
@@ -120,9 +132,9 @@ library ArrayUtil {
         if (z >= x.length) return x;
         y = new uint256[](x.length - 1);
         for (uint256 i; i < x.length; ) {
+            unchecked {
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:135]: Lets not wrap everything in `unchecked`

File: src/libraries/ArrayUtil.sol [hunk 2]
```diff
@@ -136,9 +148,9 @@ library ArrayUtil {
         if (z >= x.length) return x;
         y = new int256[](x.length - 1);
         for (uint256 i; i < x.length; ) {
+        unchecked {
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:151]: same as above

File: src/libraries/ArrayUtil.sol [hunk 3]
```diff
@@ -244,26 +256,22 @@ library ArrayUtil {
         if (left == right) return;
         int256 i = left;
         int256 j = right;
-        uint256 pivot = arr[uint256(left + (right - left) / 2)];
+        unchecked{
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:259]: same as above

File: src/libraries/ArrayUtil.sol [hunk 4]
```diff
@@ -504,12 +523,17 @@ library ArrayUtil {
         int256[] memory a,
         int256[] memory b,
         uint256 z
-    ) internal pure {
-        for (uint256 i = 0; i < a.length; i++) {
+    ) internal pure returns (int256[] memory) {
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:526]: no returns... this function edits the array in-place.

**Review by dsshap — CHANGES_REQUESTED** [review]
Couple things to update.

Let's not get too carried away by wrapping a ton of stuff in `unchecked`.

**Comment by antoncoding** [comment]
you can run yarn prettier for linting

File: src/libraries/ArrayUtil.sol (hunk 1 above)
Comment (Fletch153) [reply src/libraries/ArrayUtil.sol:135]: The if/else needs to be in the same unchecked block in order to stop i-1 from being checked. y[i-1] should never underflow

File: src/libraries/ArrayUtil.sol (hunk 2 above)
Comment (Fletch153) [reply src/libraries/ArrayUtil.sol:151]: see response above

**Comment by antoncoding** [comment]
linting is now `forge fmt` 😂

File: src/libraries/ArrayUtil.sol [hunk 5]
```diff
@@ -397,16 +380,24 @@ library ArrayUtil {
...
+                ++i;
+            }
         }
     }
 
     function append(bytes32[] memory x, bytes32 e) internal pure returns (bytes32[] memory y) {
         y = new bytes32[](x.length + 1);
         uint256 i;
-        for (i = 0; i < x.length; i++) {
+        for (i = 0; i < x.length; ) {
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:395]: @Fletch153 there are a few places that initialize `i = 0` when it's not necessary. Can you go through the methods to make sure we aren't leaking gas?

File: src/libraries/ArrayUtil.sol [hunk 6]
```diff
@@ -479,34 +470,37 @@ library ArrayUtil {
...
-    function populate(
-        int256[] memory a,
-        int256[] memory b,
-        uint256 z
-    ) internal pure {
-        for (uint256 i = 0; i < a.length; i++) {
+    function populate(int256[] memory a, int256[] memory b, uint256 z) internal pure {
+        for (uint256 i = 0; i < a.length; ) {
             a[z + i] = b[i];
+            unchecked {
```
Comment (dsshap) [inline src/libraries/ArrayUtil.sol:501]: nit picky, but can you add a space here. It seems like we are mostly putting a new line before `unchecked` blocks.

Comment (dsshap) [reply src/libraries/ArrayUtil.sol:501]: Looks like there a few places where this comments applies.

**Review by dsshap — CHANGES_REQUESTED** [review]

**Review by dsshap — APPROVED** [review]

**Pull request merged**